### PR TITLE
[Merged by Bors] - feat: generalize `DirectSum` results to avoid negation

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Three files are modified, where the hypotheses are relaxed from `Ring` or `CommRing` to `Semiring` or `CommSemiring`, and `AddCommGroup` to `AddCommMonoid`.
Besides this, no definition is changed, and for one proof in `RingTheory.Flat.Basic`, I needed to add an instance (`letI`…) in the proof.
(Everything pertains to direct sums of modules.)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by AntoineChambert-Loir** [comment]
The compilation of `RingTheory.Flat.Basic` doesn't work well. 
Lean can't find an instance that a direct sum of groups is a group…

**Comment by AntoineChambert-Loir** [comment]
So I could make it work, but this is strange.
```
noncomputable local instance (ι : Type v)
    (M : ι → Type w) [(i : ι) → AddCommGroup (M i)] [(i : ι) → Module R (M i)]
    (I : Ideal R) :
    AddCommGroup (⨁ i, ↥I ⊗[R] M i) :=
  instAddCommGroupDirectSumToAddCommMonoid fun i ↦ ↥I ⊗[R] M i
```

I found it using `apply?`, but `inferInstance` doesn't work.

**Comment by riccardobrasca** [comment]
!bench

**Comment by AntoineChambert-Loir** [comment]
!bench

**Comment by AntoineChambert-Loir** [comment]
@riccardobrasca , can you launch `bench` again? — I fear the missing dostring made it fail.

**Comment by riccardobrasca** [comment]
!bench

**Comment by riccardobrasca** [comment]
I think anyone can do it (but I am not sure)

**Comment by AntoineChambert-Loir** [comment]
(I tried, and nothing hapenned… apparently, one needs admin privileges…)

File: Mathlib/RingTheory/Flat/Basic.lean [hunk 1]
```diff
@@ -222,6 +223,7 @@ instance directSum (ι : Type v) (M : ι → Type w) [(i : ι) → AddCommGroup
     h₃, LinearMap.map_eq_zero_iff] at f
   simp [f]
 
+
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Flat/Basic.lean:226]:
```suggestion
```

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean [hunk 2]
```diff
@@ -24,8 +24,8 @@ open DirectSum
 
 open Set LinearMap Submodule
 
-variable {R : Type u} {M : Type v} {N : Type w} [Ring R] [AddCommGroup M] [Module R M]
-  [AddCommGroup N] [Module R N]
+variable {R : Type u} {M : Type v} {N : Type w} [CommSemiring R] [AddCommMonoid M] [Module R M]
+  [AddCommMonoid N] [Module R N]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:28]: This looks like it became less general in some directions? You added a commutativity requirement that wasn't here before.

File: Mathlib/RingTheory/Flat/Basic.lean [hunk 3]
```diff
@@ -187,6 +187,7 @@ instance directSum (ι : Type v) (M : ι → Type w) [(i : ι) → AddCommGroup
   classical
   rw [iff_rTensor_injective]
   intro I hI
+  letI : ∀ i, AddCommGroup (↥I ⊗[R] (M i)) := fun i ↦ TensorProduct.addCommGroup
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Flat/Basic.lean:190]: This probably deserves a TODO comment, and a link to https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2310828.20-.20generalizing.20CommRing.20to.20CommSemiring.20etc.2E/near/422684923

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:28]: Ha! This line was totally useless here, all subsequent lemmas reintroduce assumptions… (Phew…)

File: Mathlib/RingTheory/Flat/Basic.lean (hunk 3 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/Flat/Basic.lean:192]: done.

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean]: I deleted these two lines…

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**